src/hugr/typecheck.rs

```diff
@@ -0,0 +1,163 @@
+    /// against, as above, but for rows instead of simple types
+    TypeRowMismatch(TypeRow, TypeRow),
+}
+
+impl Display for TypeError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let str = match self {
+            TypeError::Unimplemented(ty) => {
+                format!("Const type checking unimplemented for {}", ty)
+            }
+            TypeError::Failed(typ) => format!("Invalid const value for type {}", typ),
+            TypeError::IntTooLarge(width, val) => {
+                format!("Const int {} too large for type I{}", val, width)
+            }
+            TypeError::IntWidthTooLarge(w) => format!("Int type too large: I{}", w),
+            TypeError::IntWidthInvalid(w) => {
+                format!(
+                    "The int type I{} is invalid, because {} is not a power of 2",
+                    w, w
+                )
+            }
+            TypeError::IntWidthMismatch(exp, act) => format!(
+                "Type mismatch for int: expected I{}, but found I{}",
+                exp, act
+            ),
+            TypeError::ConstCantBeVar => "Type of a const value can't be Var".to_string(),
+            TypeError::TupleWrongLength => "Tuple of wrong length".to_string(),
+            TypeError::LinearTypeDisallowed => {
+                "Linear types not allowed in const nodes".to_string()
+            }
+            TypeError::InvalidSumTag => "Tag of Sum value is invalid".to_string(),
+            TypeError::TypeMismatch(exp, act) => {
+                format!("Type mismatch for const - expected {}, found {}", exp, act)
+            }
+            TypeError::TypeRowMismatch(exp, act) => {
+                format!("Type mismatch for const - expected {}, found {}", exp, act)
+            }
+        };
+        f.write_str(&str)
+    }
+}
+
+/// Per the spec, valid widths for integers are 2^n for all n in [0,7]
+fn check_valid_width(width: u8) -> Result<(), TypeError> {
+    if width > 128 {
+        return Err(TypeError::IntWidthTooLarge(width));
+    }
+
+    let valid_widths: Vec<u8> = (0..8).map(|a| u8::pow(2, a)).collect();
+    if valid_widths.contains(&width) {
+        Ok(())
+    } else {
+        Err(TypeError::IntWidthInvalid(width))
+    }
+}
+
+/// Typecheck a constant value - for ensuring that it's valid before creating
+/// a const node
+pub fn typecheck_const(typ: &ClassicType, val: &ConstValue) -> Result<(), TypeError> {
+    match (typ, val) {
+        // N.B. If the width is larger than u6, our const value (backed by an i64)
+        // wont be able to accomodate the value anyway.
+        (ClassicType::Int(exp_width), ConstValue::Int { value, width }) => {
+            // Check that the types make sense
+            check_valid_width(*exp_width)?;
+            check_valid_width(*width)?;
+            // Check that the terms make sense against the types
+            if exp_width == width {
+                let max_value = u128::pow(2, *width as u32);
```

Comment:
similarly the backing type should be declared somewhere as `pub type HUGR_INT_STORE = u128;`